src/python/pants/backend/javascript/subsystems/nodejs_tool.py

```diff
@@ -0,0 +1,187 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import ClassVar, Iterable, Mapping
+
+from pants.backend.javascript import install_node_package, nodejs_project_environment
+from pants.backend.javascript.install_node_package import (
+    InstalledNodePackage,
+    InstalledNodePackageRequest,
+)
+from pants.backend.javascript.nodejs_project_environment import NodeJsProjectEnvironmentProcess
+from pants.backend.javascript.resolve import FirstPartyNodePackageResolves, NodeJSProjectResolves
+from pants.backend.javascript.subsystems.nodejs import NodeJS, NodeJSToolProcess
+from pants.engine.internals.native_engine import Digest, MergeDigests
+from pants.engine.internals.selectors import Get
+from pants.engine.process import Process
+from pants.engine.rules import Rule, collect_rules, rule
+from pants.engine.unions import UnionRule
+from pants.option.option_types import StrOption
+from pants.option.subsystem import Subsystem
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+from pants.util.strutil import softwrap
+
+
+class NodeJSToolBase(Subsystem):
+    # Subclasses must set.
+    default_version: ClassVar[str]
+
+    version = StrOption(
+        advanced=True,
+        default=lambda cls: cls.default_version,
+        help="Version string for the tool in the form package@version (e.g. prettier@2.6.2)",
+    )
+
+    install_from_resolve = StrOption(
+        advanced=True,
+        default=None,
+        help=lambda cls: softwrap(
+            f"""\
+            If specified, install the tool using the lockfile for this named resolve,
+            instead of the version configured in this subsystem.
+
+            If unspecified, the tool will use the default configured package manager
+            [{NodeJS.options_scope}].package_manager`, and install the tool without a
+            lockfile.
+            """
+        ),
+    )
```

Comment:
When this flag is used, is the `version` automatically included in the inputs to the relevant resolve, or does it need to be repeated in those inputs? @benjyw has been making recent adjustments in that area on the Python side.

Reply:
I second that question. As of 2.16.0rc3, in the PythonToolBase, you use `requirements=` to either specify the requirements needed from the lockfile, or, preferably, addresses of python_requirement targets that specify them, to avoid repetition.

"requirements" is I guess the wrong word to use in the JS world, but what is the right word?  For consistency we should probably replace `version` with that. 

However, this is all only relevant if we're able to A) subset a JS lockfile, or B) want to validate that the lockfile provides the right things. Are either of those relevant here?

Right now it looks like `version=` lets you bypass using a lockfile? Not sure we need to support that?

Reply:
Subsetting is not a thing currently, no.

I think `package` is the term used in JS land.

I think it is rather that `install_from_resolve` bypasses the configured 
default version. Setting a version and resolve means the tool will be ran with the resolve, the version is ignored.

As for supporting running tools without lockfile. That is already an in use feature, I dont think this is the pr to remove it.

Reply:
Maybe not in this PR, but in python-land we are moving away from supporting installing tools with no lockfile. And it would simplify things here to if there were only one way to customize a tool version, which is via lockfile. We could then get rid of `version` here, IIUC. So you either use a built-in lockfile, or a custom one (set by `install_from_resolve`). That seems better than having this confusing state where one option causes another to be ignored.